src/SNLS_HybrdTrDLDenseG.h

```diff
@@ -0,0 +1,599 @@
+        if ( m_outputLevel > 1 && m_os != nullptr ) {
+            *m_os << "Function and Jacobian evaluations: " << m_fevals << " " << m_jevals << std::endl;
+        }
+#endif
+    };
+                
+    __snls_hdev__ 
+    int getNDim() const { return m_nDim; }
+    __snls_hdev__ 
+    int getNFEvals() const { return m_fevals; }
+    __snls_hdev__ 
+    int getNJEvals() const { return m_jevals; }
+    __snls_hdev__ 
+    double getDelta() const { return m_delta; }
+    __snls_hdev__ 
+    double getRes() const { return m_res; }
+
+    // setX can be used to set the initial guess
+    __snls_hdev__ 
+    inline 
+    void setX( const double* const x ) 
+    {
+        for (int iX = 0; iX < m_nDim; ++iX) {
+            m_x[iX] = x[iX] ;
+        }
+    };
+
+    __snls_hdev__ 
+    inline 
+    void getX( double* const x ) const 
+    {
+        for (int iX = 0; iX < m_nDim; ++iX) {
+            x[iX] = m_x[iX] ;
+        }
+    };   
+
+    /**
+     * Must call setupSolver before calling solve
+    */
+    __snls_hdev__ 
+    inline
+    void setupSolver(int maxIter,
+                     double tolerance,
+                     TrDeltaControl * deltaControl,
+                     int outputLevel = 0) 
+    {
+        m_status = unConverged;
+        m_fevals = 0;
+        m_jevals = 0;
+
+        m_maxIter = maxIter;
+        m_maxfev = maxIter;
+        m_tolerance = tolerance;
+        m_deltaControl = deltaControl;
+
+        this->setOutputlevel( outputLevel );
+
+        m_complete = true;
+
+    }
+
+    __snls_hdev__ 
+    void setOutputlevel( int outputLevel ) {
+        m_outputLevel = outputLevel;
+        m_os          = nullptr;
+        //
+#ifdef __cuda_host_only__
+        if ( m_outputLevel > 0 ) {
+            m_os = &(std::cout);
+        }
+#endif
+    }
+
+    // solve returns status
+    //
+    // on exit, _res is consistent with _x
+    __snls_hdev__ 
+    inline
+    SNLSStatus_t solve() {
+    
+        if ( !m_complete) {
+        SNLS_FAIL("solve", "Setup not called beforehand!") ;
+        }
+
+        double residual[m_nDim];
+        double Jacobian[m_nXnDim];
+
+        // Working arrays
+        double Qmat[m_nXnDim];
+        double work_arr1[m_nDim];
+        double work_arr2[m_nDim];
+        double work_arr3[m_nDim];
+        double qtf[m_nDim];
+        // Do initial solver checks
+        m_status = solveInit(residual);
+        if (m_status != unConverged) {
+            return m_status;
+        }
+
+        // Run our solver until it converges or fails
+        while (m_status == unConverged && m_nIters < m_maxIter && m_fevals < m_maxIter) {
+            solveStep(residual, Jacobian, Qmat, qtf, work_arr1, work_arr2, work_arr3);
+        }
+
+        return m_status;
+    }
+
+    // convenience wrapper, for the current _x
+    __snls_hdev__ 
+    inline bool computeRJ(double* const r,
+                          double* const J ) 
+    {
+        m_fevals += ((J == nullptr) ? 1 : 0);
```

Comment:
So, this does act the same in principal. The only time we call `computeRJ` in this solver with `J != nullptr` is when we need to update the Jacobian. We actually don't care about the residual outputted at that point in time because we already have it from either `solveInit` or from within the `for(;;)` portion of `solveStep`.

Reply:
Understood. I guess it comes down to what we want fevals to track -- is it the number of times the function evaluation happens or the number of times that the solver needs the function evaluation. The computeRJ is set up to always compute the residual and to optionally compute the Jacobian. And I was interpreting fevals as the number of evaluations rather than the number of times an evaluation is actually needed. So in that meaning of fevals I think it would get incremented with every computeRJ call.

Reply:
Okay if that's how you're interpreting things than I can have it increment every time.